math/src/fft/fft_inputs.rs

```diff
@@ -69,17 +104,98 @@ pub trait FftInputs<B: StarkField> {
     ///
     /// # Panics
     /// Panics if length of the `twiddles` parameter is not self.len() / 2.
-    fn fft_in_place(&mut self, twiddles: &[B]) {
+    fn fft_in_place(&mut self, twiddles: &[E::BaseField]) {
         fft_in_place(self, twiddles, 1, 1, 0);
     }
+
+    // CONCURRENT METHODS
+    // --------------------------------------------------------------------------------------------
+
+    // PERMUTATIONS
+    // ================================================================================================
+
+    #[cfg(feature = "concurrent")]
+    /// Permutes the elements in this input using the permutation defined by the given
+    /// permutation index in a concurrent manner.
+    fn permute_concurrent(&mut self)
+    where
+        Self: Send,
+    {
+        let n = self.len();
+        let num_batches = rayon::current_num_threads().next_power_of_two();
+        let batch_size = n / num_batches.min(n);
+        rayon::scope(|s| {
+            for batch_idx in 0..num_batches {
+                // create another mutable reference to the slice of values to use in a new thread; this
+                // is OK because we never write the same positions in the slice from different threads
+                let values = unsafe { &mut *(&mut *self as *mut Self) };
```

Comment:
We should do that only if the feature `concurrent` is enabled. Otherwise, it should `let values = self;` as there is no reason to do so if there is no parallelization.

Also, efore `unsafe` blocks, it's good to prefix comments with a `// Safety: ___` as in the convention

Fianlly, one alternative way to write that, that would be slightly more readable, is:

```rust
#[cfg(feature = "concurrent")]
// Safety: we bypass the lifetime check in order to allow parallelization as the
// algorithm is guaranteed to not have overlapping mutations.
let values = unsafe { (self as *mutSelf).as_mut().unwrap_unchecked() };

#[cfg(not(feature = "concurrent"))]
let values = self;
```